swupd/delta.go

```diff
@@ -0,0 +1,177 @@
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/pkg/errors"
+)
+
+// CreateDeltas creates all delta files between the previous and current
+// version of the supplied manifest, returning a list of files that
+// delta creation failed for.
+func CreateDeltas(manifest, statedir string, from, to uint32) ([]*File, error) {
+	var c config
+
+	c, err := getConfig(statedir)
+	if err != nil {
+		return nil, err
+	}
+
+	var oldManifest *Manifest
+	var newManifest *Manifest
+
+	if oldManifest, err = ParseManifestFile(filepath.Join(c.outputDir, fmt.Sprintf("%d", from), manifest)); err != nil {
+		return nil, err
+	}
+	if newManifest, err = ParseManifestFile(filepath.Join(c.outputDir, fmt.Sprintf("%d", to), manifest)); err != nil {
+		return nil, err
+	}
+
+	// Must be sorted before passing into linkPeersAndChange()
+	oldManifest.sortFilesName()
+	newManifest.sortFilesName()
+
+	_, _, _ = newManifest.linkPeersAndChange(oldManifest, from)
+
+	deltas, err := consolidateDeltaFiles(newManifest, from, c)
```

Comment:
Nitpick:

You aren't being consistent about how you are assigning to `err`. First you do
```go
c, err := getConfig(statedir)
```
so at this point err is declared and in scope. Then you do
```go
if oldManifest, err = ParseManifestFile(...
```
so you expect it to be in scope because of the missing `:`. But then here you are shadowing `err` to declare a new one with `:=`

Reply:
@matthewrsj Not true. The err here is the one declared at line 19. The `:=` has to declare at least one new variable, but it doesn't declare all of them to be new.  So this line creates a new variable `deltas` but reuses the existing err as it is in the same lexical block. See [playground](https://play.golang.org/p/AP9XSvOz4VX) for a poor example. See section 2.3.1 of the Go Programming Language, starting "One subtle but important point" (page 31 in my copy.

Reply:
Ah thanks @icarus-sparry, TIL.